src/metabase/driver/common/parameters/dates.clj

```diff
@@ -201,36 +202,80 @@
 (def ^:private all-date-string-decoders
   (concat relative-date-string-decoders absolute-date-string-decoders))
 
-(defn- execute-decoders
+(s/defn ^:private execute-decoders
   "Returns the first successfully decoded value, run through both parser and a range/filter decoder depending on
-  `decoder-type`."
-  [decoders decoder-type decoder-param date-string]
+  `decoder-type`. This generates an *inclusive* range by default. The range is adjusted to be exclusive as needed: see
+  dox for `date-string->range` for more details."
+  [decoders, decoder-type :- (s/enum :range :filter), decoder-param, date-string :- s/Str]
   (some (fn [{parser :parser, parser-result-decoder decoder-type}]
           (when-let [parser-result (parser date-string)]
             (parser-result-decoder parser-result decoder-param)))
         decoders))
 
-(s/defn date-string->range :- {(s/optional-key :start) s/Str, (s/optional-key :end) s/Str}
+(def ^:private TemporalRange
+  {(s/optional-key :start) Temporal, (s/optional-key :end) Temporal})
+
+(s/defn ^:private adjust-inclusive-range-if-needed :- (s/maybe TemporalRange)
+  "Make an inclusive date range exclusive as needed."
+  [{:keys [inclusive-start? inclusive-end?]}, {:keys [start end], :as m} :- (s/maybe TemporalRange)]
+  (merge
+   (when start
+     {:start (if inclusive-start?
+               start
+               (u.date/add start :day -1))})
+   (when end
+     {:end (if inclusive-end?
+             end
+             (u.date/add end :day 1))})))
+
+(def ^:private DateStringRange
```

Comment:
added extra schemas and documentation to clarify what everything here is supposed to be doing